### PR TITLE
let users refresh mutuals via button with rate limiting

### LINKED ISSUE #14: button for updating mutuals

### PR DESCRIPTION
**Pull request by morgangallant**
fixes #14

### CODE REVIEW

**Comment by morgangallant** [comment]
and fixes #12

File: src/firebase.js [hunk 1]
```diff
@@ -30,26 +30,51 @@ export const signInWithTwitter = () => {
...
+
+export const updateTwitterMutuals = async () => {
+  var user = auth.currentUser;
+  if (!user) {
+    return;
+  }
+  const profile = await fetchProfile(user.uid);
+  // Check if the user has recently fetched mutuals. If they have done so in the
+  // last five minutes, an error will be shown to them.
+  const lastfetch = (profile.lastMutualFetch) ? profile.lastMutualFetch : new Date(0);
```
Comment (kipply) [inline src/firebase.js:43]: can you also write a firestore rule for this? 

you can do it with data validation (ensure the value being written is equal to / close to the current time)

File: src/firebase.js [hunk 2]
```diff
@@ -30,26 +30,51 @@ export const signInWithTwitter = () => {
...
+}
+
 export const updateUser = async result => {
   firebase.firestore().collection(`users/`).doc(result.user.uid).set({
     tuid: result.additionalUserInfo.profile.id_str,
     handle: result.additionalUserInfo.profile.screen_name,
     displayName: result.user.displayName,
     profilePicture: result.additionalUserInfo.profile.profile_image_url.replace("_normal", ""),
+    // according to stack overflow, this is ok.
+    ttoken: result.credential.accessToken,
```
Comment (kipply) [inline src/firebase.js:74]: uhh let's read protect this, which will involve nesting this into `users/{userID}/credentials/` (we can only make firestore read rules for collections, not fields)

**Comment by morgangallant** [comment]
should be good now

File: firestore.rules [hunk 3]
```diff
@@ -28,5 +28,8 @@ service cloud.firestore {
         get(/databases/$(database)/documents/users/$(userId)/likes/$(userId2)).data.like == true &&
         get(/databases/$(database)/documents/users/$(userId2)/likes/$(userId)).data.like == true
     }
+    match /users/{userId}/credentials {
```
Comment (kipply) [inline firestore.rules:31]: data validation for the last mutual fetch?

**Comment by morgangallant** [comment]
Probably something I’m missing, could you explain what you mean by this? Seems a bit odd to me that we would need to add a database rule to validate something that we already protect against in our software.
"vercel[bot]" <notifications@github.com> wrote:
“@kipply commented on this pull request.”


“In firestore.rules:”

> @@ -28,5 +28,8 @@ service cloud.firestore {
         get(/databases/$(database)/documents/users/$(userId)/likes/$(userId2)).data.like == true &&
         get(/databases/$(database)/documents/users/$(userId2)/likes/$(userId)).data.like == true
     }
+    match /users/{userId}/credentials {
“data validation for the last mutual fetch?”

File: src/firebase.js (hunk 1 above)
Comment (kipply) [reply src/firebase.js:43]: you add two new fields in this pr. the tokens are protected by read/write rules, as is the `lastMutualFetch`. 

It is not secure in that the user can still write arbitrary values to `lastmutualfetch`, making the throttle useless. Thus, data validation on top of access-security is necessary for the database rules

**Comment by morgangallant** [comment]
Got it - will fix tonight.

**Pull request merged**